CA-226886: VDI import: handle exn if SR is unavailable for any host

minli1 opened the pull request:
CA-226886: VDI import: handle exn if SR is unavailable for any host.

This commit adds proper handling in VDI import for exceptions raised
when the SR is not accessible by any host.  For example, this can
happen during an update-upload operation when trying to use the
default SR but it was down.

The exception handler marks the relevant task as failed, and returns
an appropriate http error code to the client.

gaborigloi commented on ocaml/xapi/import_raw_vdi.ml:
```diff
@@ -114,13 +114,24 @@ let import vdi (req: Request.t) (s: Unix.file_descr) _ =
+                (fun __context ->
+                  TaskHelper.failed ~__context Api_errors.(Server_error(vdi_missing,[]));
+                  Http_svr.headers s (Http.http_400_badrequest ())
+                );
```
I think it would be better to use simpler helper functions to simply extract the task from the query, instead of calling `Xapi_http.with_context`, as that function does many things that we probably don't need here. I think we just need to get the task, set it to failed, and return an HTTP error code.

minli1 replied:
Yes, will do it. My initial thought of this change is to get the task_id from the request.
Thanks

gaborigloi commented on ocaml/xapi/import_raw_vdi.ml:
```diff
@@ -114,13 +114,24 @@ let import vdi (req: Request.t) (s: Unix.file_descr) _ =
+                      TaskHelper.failed ~__context Api_errors.(Server_error(no_hosts_available,[]));
+                      Http_svr.headers s (Http.http_400_badrequest ());
+                    )
+                end;
```
I think it's easier if we handle this exception in a try/with block around this function. Then we don't have to modify `find_host_for_sr`, we can just catch the exception it throws (`Api_errors.no_hosts_available`). Also, I think that we should catch any exception that occurs here, and set the task to failed, and return an HTTP error code.

minli1 replied:
Refined.

gaborigloi commented on ocaml/xapi/importexport.ml:
```diff
@@ -188,7 +188,9 @@ let check_sr_availability ~__context sr =
-  Xapi_vm_helpers.choose_host ~__context ~choose_fn ~prefer_slaves ()
+  try  
+    Xapi_vm_helpers.choose_host ~__context ~choose_fn ~prefer_slaves ()
+  with _ -> Ref.null
```
Please see the first, related comment - it's best not to modify this function and catch the exception it throws instead, because this function is called from multiple places. Therefore it's best not to change the behaviour of this function because we don't know how it will affect the other callers.

minli1 replied:
Yes, agree with it, will do. Thanks

gaborigloi commented:
Thanks for the PR, I think the overall strategy looks good, this will finally fix the underlying cause of CA-253489 and CA-226886. We just need to make some changes to the implementation that I've described in my comments.

thomassa commented:
It looks as if this will be a good code-change once @gaborigloi's suggestions have been incorporated.
Please would you change the commit message as well, though?
It should start
`CA-226886: ` rather than
`ca-226886 :` (note the capitalisation, and the ordering of the space and colon characters.)

and generally we like to make a commit-message describe the change being made in the commit, so in this case it might be
```
CA-226886: VDI import: handle excn if SR is unavailable

This commit adds proper handling in VDI import for exceptions raised
when the SR is not accessible by any host.  For example, this can
happen during an update-upload operation when trying to use the
default SR if there is no default SR set.

The exception handler marks the relevant task as failed, and returns
an appropriate http error code to the client.
```

Also, when we merge a Github pull-request, Github takes the pull-request's headline description and first comment and uses them as the commit-message for the merge-commit.
I don't think you want "Hi, all. Please help to review the code." appearing in the git log!

minli1 commented:
Hi, @thomassa  Thanks for your comments, it's really helpful for me. I had changed the title and the commit-message.  : )

minli1 commented:
Hi, @gaborigloi  Could you please have a look again since I had refine codes according to the comments.
Please help to review and merge if possible. Thanks  : )

mseri commented:
@minli1 I see the pull request comment updated by the commit messages and titles do not seem updated, also the new commit is not following @thomassa requested style.

gaborigloi commented on ocaml/xapi/import_raw_vdi.ml:
```diff
@@ -112,17 +112,45 @@ let import vdi (req: Request.t) (s: Unix.file_descr) _ =
-                None
+              else 
+                begin
+                  try 
```
I think we could wrap even more code with this try/with block, to catch all errors in the handler. It might be better to wrap the whole import function, to catch any errors that occur in it, by moving this try/with block to wrap lines 161-166 in `let handler ...`.

mseri commented on ocaml/xapi/import_raw_vdi.ml:
```diff
@@ -112,17 +112,45 @@ let import vdi (req: Request.t) (s: Unix.file_descr) _ =
+                    error "Caught exception in import handler: %s" (ExnHelper.string_of_exn e);
+                    let all = req.Request.cookie @ req.Request.query in 
+                    let task_id = 
+                    if List.mem_assoc "task_id" all 
```
These three lines should be indented more. Can you run `ocp-indent` on the modified files before committing the changes please?

minli1 replied:
Sure, will run it. Thanks.

mseri commented on ocaml/xapi/import_raw_vdi.ml:
```diff
@@ -112,17 +112,45 @@ let import vdi (req: Request.t) (s: Unix.file_descr) _ =
+              let e = Api_errors.(Server_error(vdi_missing,[])) in 
+              let all = req.Request.cookie @ req.Request.query in
+              let task_id =
+              if List.mem_assoc "task_id" all 
```
As above.

gaborigloi commented on ocaml/xapi/import_raw_vdi.ml:
```diff
@@ -112,17 +112,45 @@ let import vdi (req: Request.t) (s: Unix.file_descr) _ =
+              then Some (Ref.of_string (List.assoc "task_id" all))
+              else None in
+              begin match task_id with
+              | None -> Server_helpers.exec_with_new_task "import" ~task_in_database:true (fun __context -> TaskHelper.failed ~__context e)
```
If I'm not mistaken, we do not need this line, because it will set a newly-created task to failed, which is not useful for the caller?
Oh, or I assume you created this task to indicate that something failed. In that case, I think we don't need to create a new task here, maybe we could just use the context from line 101: ` Server_helpers.exec_with_new_task "VDI.import" ...` and set that to failed.

minli1 replied:
Two concern about this:
1. When will we get the situation of task_id is None? I think we can always get a task_id and the None case will never get into it.  
See here : https://github.com/xapi-project/xen-api/blob/master/ocaml/xapi/cli_util.ml#L99
2. If we got a task id None, we set the context we just created in Line 101, it seems that we still cannot finish the task from the caller.

So, I think if we believe in the task_id will not be None, shall we remove the None case ? Thanks

gaborigloi replied:
Yeah, actually I think it's best to unconditionally  set the other, locally-created context from line 101 to failed, regardless of whether `task_id` is in the request or not, because that would provide some useful information in the logs - we would see in the logs that this `VDI.import` task that we started has failed.

minli1 replied:
Yes, agreed. I will fail the task with the current task. Thanks .

gaborigloi commented on ocaml/xapi/import_raw_vdi.ml:
```diff
@@ -112,17 +112,45 @@ let import vdi (req: Request.t) (s: Unix.file_descr) _ =
+                    let task_id = 
+                    if List.mem_assoc "task_id" all 
+                    then Some (Ref.of_string (List.assoc "task_id" all))
+                    else None in
```
I think this is a good opportunity to factor out some duplication :). If you check `Xapi_http.assert_credentials_ok` and `Xapi_http.with_context`, the code getting the `task_id` and `subtask_of` from the request is duplicated between them. I think these could be extracted into `task_id_of_req` and `subtask_of_of_req` helpers in `importexport.ml`.

minli1 replied:
Thanks ,will refactor some of the codes to make it simpler.

gaborigloi commented on ocaml/xapi/import_raw_vdi.ml:
```diff
@@ -112,17 +112,45 @@ let import vdi (req: Request.t) (s: Unix.file_descr) _ =
+              else None in
+              begin match task_id with
+              | None -> Server_helpers.exec_with_new_task "import" ~task_in_database:true (fun __context -> TaskHelper.failed ~__context e)
+              | Some task_id -> Server_helpers.exec_with_forwarded_task task_id (fun __context -> TaskHelper.failed ~__context e)
```
I think `Xapi_http.assert_credentials_ok` does this in a simpler way: https://github.com/xapi-project/xen-api/blob/master/ocaml/xapi/xapi_http.ml#L104 , we could do the same thing here:
```ocaml
| Some task_id -> TaskHelper.failed ~__context:(Context.from_forwarded_task task_id) e
```

minli1 replied:
Yes, agreed.
Will change it to a simple way. Thanks.

gaborigloi replied:
We could even use our `stdext` library here, like this:
```ocaml
Stdext.Opt.map (fun task_id -> TaskHelper.failed ~__context:(Context.from_forwarded_task task_id) e) task_id
```

thomassa commented:
@minli1, you haven't changed the commit-message on the commit itself. (To do that you can use `git rebase -i 9090ca19ff30c9b8e3014ce082ec7eec4f962dee` and choose "reword" for the relevant commit, and then afterwards push the changes to your github repository.)

minli1 commented:
Hi, @gaborigloi @mseri  @thomassa 
Could you please help to review the refined codes?
Thanks a lot.

minli1 commented on ocaml/xapi/import_raw_vdi.ml:
```diff
@@ -98,33 +98,51 @@ let import vdi (req: Request.t) (s: Unix.file_descr) _ =
+    let task_id = Xapi_http.task_id_of_req req in
+    begin match task_id with 
+      | Some task_id -> TaskHelper.failed ~__context:(Context.from_forwarded_task task_id) e 
+      | None -> Server_helpers.exec_with_new_task "import" ~task_in_database:true (fun __context -> TaskHelper.failed ~__context e)
```
@gaborigloi  Hi, Gabor. I wrap almost all the import function with try/with according to your comment.
However if wrap all, I could not get a ~__context so I have to failed the task by creating a new one, not like the L127.
The purpuse for this is to raise a exn here, I think it is same as I set that task failed in L102. 
It's just help us to know where is the problem. Also I think task_id=None will never comes up unless something wrong.
What do you think of this code change? Or there is some better way. Thanks a lot.

gaborigloi replied:
Hi @minli1, I agree, I think we should have the try/with inside
```
Server_helpers.exec_with_new_task "VDI.import"
      (fun __context ->
```
instead of wrapping almost the same function.

minli1 replied:
: ) 
I will change it. Thanks

gaborigloi replied:
Yes, I think we don't need to deal with the `None` case, so we can simplify it even more using the "piping" function `|>`:
```ocaml
Xapi_http.task_id_of_req req |> Stdext.Opt.map (fun task_id -> TaskHelper.failed ~__context:(Context.from_forwarded_task task_id) e)
```
Or even:
```ocaml
req |> Xapi_http.task_id_of_req |> Stdext.Opt.map (fun task_id -> TaskHelper.failed ~__context:(Context.from_forwarded_task task_id) e)
```

minli1 replied:
Sure, I had changed it to a simple way. Thanks.

gaborigloi commented on ocaml/xapi/xapi_http.ml:
```diff
@@ -316,3 +316,8 @@ let add_handler (name, handler) =
+    let all = Http.Request.(req.cookie @ req.query) in
+    if List.mem_assoc "task_id" all
+    then Some (Ref.of_string (List.assoc "task_id" all))
+    else None
```
The second, refactoring commit could also use this inside `Xapi_http.assert_credentials_ok` and `Xapi_http.with_context`.

gaborigloi commented on ocaml/xapi/import_raw_vdi.ml:
```diff
@@ -98,33 +98,51 @@ let import vdi (req: Request.t) (s: Unix.file_descr) _ =
+    None
+  | e ->
+    error "Caught exception in import handler : %s" (Printexc.to_string e);
+    raise e    
```
I think it would be good to return an HTTP error and set the task to failed also in this case, where we catch all exceptions. Otherwise the client is never notified, if an error other than `Api_errors.Server_error` happens, and can still hang causing issues like CA-226886 and CA-253489, if I'm not mistaken?

thomassa replied:
I agree.

minli1 replied:
Hi, thanks all. I had changed it to  with e -> ... ;
So that all exn could be catched and make the original task failed.

gaborigloi commented:
Oh, and another things: this is a bit subjective, but I think the commit message of the second commit could be changed to indicate how it's different from the first commit, that it's factoring out duplication.

thomassa commented on ocaml/xapi/import_raw_vdi.ml:
```diff
@@ -98,33 +98,51 @@ let import vdi (req: Request.t) (s: Unix.file_descr) _ =
          )
     )
+  with 
+  | Api_errors.Server_error (a,b) as e ->
```
Since `a` and `b` are never used, this could be
`Api_errors.Server_error (_,_)` or even just
`Api_errors.Server_error _`

minli1 commented:
Hi, @gaborigloi  and @thomassa.  : )
Could please help to review the refined codes and merge if there are no other comments?
Thanks in advance.

gaborigloi commented on ocaml/xapi/import_raw_vdi.ml:
```diff
@@ -111,20 +112,27 @@ let import vdi (req: Request.t) (s: Unix.file_descr) _ =
-              else
+              then localhost_handler rpc session_id vdi req s  
+              else begin
+              try   
```
I think it's best to wrap everything inside `Server_helpers.exec_with_new_task "VDI.import"` with this try/with block, between from line 104. This would make it even safer, because I think there are still some operations outside this try/with block that can fail. So we could write it like this:
```ocaml
   Server_helpers.exec_with_new_task "VDI.import"
     (fun __context ->
         try
           Helpers.call_api_functions ~__context
             (
                ...
             )
           )
         with
             ... error handling
     )
```
This way, we still have the context so we can mark the newly-created task as failed, but we protect as much code as possible with the try/with block.

gaborigloi commented on ocaml/xapi/import_raw_vdi.ml:
```diff
@@ -111,20 +112,27 @@ let import vdi (req: Request.t) (s: Unix.file_descr) _ =
               error "Require an SR or VDI to import";
-              TaskHelper.failed ~__context Api_errors.(Server_error(vdi_missing,[]));
+              let e = Api_errors.(Server_error(vdi_missing,[])) in 
+              ignore(Xapi_http.task_id_of_req req |> Stdext.Opt.map (fun task_id -> TaskHelper.failed ~__context:(Context.from_forwarded_task task_id) e));
```
Since we are using this expression at two places, we could extract it into a local `fail_task_in_request e` helper function inside this `import` function (where `e` is our exception).

gaborigloi replied:
Actually, even better, we could also include the `Http_svr.headers s (Http.http_400_badrequest ());` part in that function, to make it even shorter, and remove even more duplication :smile:, so it could be called `fail_request` or something.

gaborigloi commented on ocaml/xapi/import_raw_vdi.ml:
```diff
@@ -111,20 +112,27 @@ let import vdi (req: Request.t) (s: Unix.file_descr) _ =
-
          )
-    )
+    )   
```
It seems that accidentally some trailing whitespace got into the commit

minli1 commented:
HI, @gaborigloi  Help to review again.  : ) 
Appreciate !

gaborigloi commented:
It's perfect now, just one thing: there is still some trailing whitespace sadly in the file. However, if you're on Linux, you can run the following command to get rid of it:
```
sed --in-place 's/[[:space:]]\+$//' ocaml/xapi/import_raw_vdi.ml
```
Also, now the second commit contains changes that are different from the refactoring. Therefore I think it's the easiest to simply squash the two PRs, mention the refactoring as well in the commit message, and remove the trailing whitespace.

After that I'll quickly test it against the bug I worked on (CA-253489) to verify that it solves the root cause of that issue too (I assume you've tested it already against your one, CA-226886), and they I'll approve :smile: .

gaborigloi commented on ocaml/xapi/import_raw_vdi.ml:
```diff
@@ -98,32 +101,38 @@ let import vdi (req: Request.t) (s: Unix.file_descr) _ =
+              | None ->
+                error "Require an SR or VDI to import";
+                let e = Api_errors.(Server_error(vdi_missing,[])) in 
+                fail_task_in_request req s e;
```
This is completely optional, but these two lines can be written in one line,
```ocaml
fail_task_in_request req s Api_errors.(Server_error(vdi_missing,[]))
```

gaborigloi commented on ocaml/xapi/import_raw_vdi.ml:
```diff
@@ -98,32 +101,38 @@ let import vdi (req: Request.t) (s: Unix.file_descr) _ =
+       with e ->
+         error "Caught exception in import handler: %s" (ExnHelper.string_of_exn e);
+         fail_task_in_request req s e;
+         None
```
I think we should reraise the exception `e` here instead of returning `None`, to preserve the old behaviour for the other callers of this function.

thomassa replied:
Agreed.
I think it is good to set the task to failed status, and to return an http error.
But then we should still raise the exception.

minli1 commented:
Hi, @gaborigloi  and @thomassa  
1. I had changed the behaviour to the old one so that it will raise exn when got an error;
2. Had run the command to avoid the trailing whitespace;
3. Changed 2 commits into one, add commit message for factor out the duplication;
4.I had tested it already and CA-226886 resolved; you can use this commit to verify if it can solve the CA-25389.
Please help to merge if there is no other comments. Thanks.

gaborigloi commented on ocaml/xapi/import_raw_vdi.ml:
```diff
@@ -98,32 +101,37 @@ let import vdi (req: Request.t) (s: Unix.file_descr) _ =
+           )
+       with e ->
+         error "Caught exception in import handler: %s" (ExnHelper.string_of_exn e);
+         fail_task_in_request req s e;
```
We don't have to change anything, just a note for the future: when a new version of `xen-api-libs-transitional` is realeased, I might be better to use the HTTP `500 Internal Server Error` error code here introduced in https://github.com/xapi-project/xen-api-libs-transitional/pull/34 instead of `400 Bad Request`.

thomassa commented on ocaml/xapi/xapi_http.ml:
```diff
@@ -44,6 +44,18 @@ open Http
+  then Some (Ref.of_string (List.assoc "task_id" all))
+  else None
+
+let subtask_of_req (req: Http.Request.t) =
```
This function-name is a bit confusing.
The meaning is really "subtask_of of req", which means the req is (or corresponds to) a subtask of the task which this function returns.
"subtask_of_of_req" would be a confusing name too, though.
So would "subtaskof_of_req" even though it might be slightly clearer.
How about "parent_task_of_req" or perhaps better still, do something like the code below?
```
let ref_param_of_req req param_name =
  let all = Http.Request.(req.cookie @ req.query) in
  try Some (Ref.of_string (List.assoc "subtask_of" all))
  with Not_found -> None
```
then use that function in
`let subtask_of = ref_param_of_req req "subtask_of" in` and
`let task_id = ref_param_of_req req "task_id" in`

minli1 replied:
Thanks.@thomassa
Agree with you, I think `let ref_param_of_req req param_name` is better one, it will reduce some duplication and is fit for other `params `in `req`.
I will change it. thanks.

minli1 replied:
Could you please help to merge if there are no other comments? Thanks. @thomassa @gaborigloi

thomassa commented:
I suggested the `try/with` instead of `List.mem` mostly because it means less code to read; also it should be slightly more efficient because the assoc-list is searched only once.
